clients/lsp-clangd.el

```diff
@@ -178,6 +178,49 @@ This must be set only once after loading the clang client.")
   :risky t
   :type '(repeat string))
 
+(defun lsp-clangd-find-compile-commands-dir ()
+  "Return the full directory path (or nil) that contain a compile_commands.json.
+
+If multiple compile_commands.json files are
+found in the project directory tree - ask the user to choose."
+  ;; TODO lsp-workspace-root won't find anything at the start of the workspace
+  (let ((candidate-dirs (directory-files-recursively (lsp-workspace-root) "compile_commands.json")))
+    (pcase (length candidate-dirs)
+          ;; TODO can I lsp--warn-user-in-message-and-log at the same time
+      (`0 (lsp--warn "Cannot find compile_commands.json - intellisense might not work")
+          nil)
+      ;; if you find 1 only - assume it's correct and strip the dirname from the full path
+      ;; happy path
+      (`1 (file-name-directory (car candidate-dirs)))
+      ;; if you have several - ask the user to choose with lsp--completing-read
+      ;; TODO if Ctrl-G from inside this completing read still falls to default lsp-clients-clangd-args
+      (_ (file-name-directory (lsp--completing-read
```

Comment:
does it make sense to prefer compile commands in the root of the project?

Reply:
my heuristic would be to start by most recently mtime. If people build out of source, compile_commands.json will be generated in the cmake `build/` directory and some may copy it to project-root. However, if compile_commands.json changes and the user forgets to copy the new file to project-root again - the UX will be suboptimal. Symlinks work sometimes, but not in docker on WSL